### DISCUSSION TITLE
How to forward e.g. progress bars sent to stdout of the process to sys.stdout?

### DISCUSSION THREAD

**Discussion by sla-te** [amoffat/sh#704] [Q&A] [answered]
```python
sh.scp(
    "-i",
    "my_key",
    "a_file.tar.gz",
    "root@192.168.1.1:/target/folder",
    _out=sys.stdout, _err=sys.stderr,
)
```

This will block until the file transfer is complete and not print anything.

In case I add the special kwargs ` _unify_ttys=True, _tty_out=True, _tty_in=True` the progress bar is printed once at 100% after the transfer is complete.

If I try to get hacky and try to adapt to https://sh.readthedocs.io/en/latest/tutorials/interacting_with_processes.html#tutorial2 like this:
```python
aggregated = ""
aggregated_stdout = ""
def print_stdout(char, stdin):
    global aggregated_stdout
    aggregated_stdout += char

    if char == "\n" or "/s" in aggregated_stdout:
        print(aggregated_stdout, end="")
        aggregated_stdout = ""

aggregated_stderr = ""
def print_stderr(char, stdin):
    global aggregated_stderr
    aggregated_stderr += char
    print(char, end="", file=sys.stderr)

sh.scp(
    "-i",
    "my_key",
    "a_file.tar.gz",
    "root@192.168.1.1:/target/folder",
    _unify_ttys=True, _out=print_stdout, _err=print_stderr,
    _out_bufsize=0, _tty_in=True,, _tty_out=True, _tee=("out", "err")
)
```
Then I am pretty much getting what I want to see (the progress bar updating) but isnt there handling for such a scenario in sh?

**Comment by amoffat** [comment]
If you don't care about capturing the output, just use `_fg=True` and let the process write whatever it wants to the terminal. Otherwise, depending on how fancy the progress bar is, you'll need to process any special ascii characters yourself to render progress correctly. Programs that do progress bars typically write all kinds of invisible characters to jump around and re-render a progress bar.

**Reply by sla-te** [reply]
Thank you for the quick and thorough reply once more, very much appreciated!
As I am wrapping the sh command via composition in my own class, that also servers logging and some other purposes, I do care about `return value` and `Exceptions`, sh raises as I use them for error handling - its important for me to have something raised if any command fails or drops warnings. From what I can tell scps progress bar is not really fancy, it just updates the bytes/second and % part. Also wouldn't like to lose  most of shs functionality. So I got it correctly, that such scenarios will need a "special" implementation like the one I created above and, that I did not over-complicate or re-invent the wheel here?

**Reply by ecederstrand** [reply]
Console progress bars are usually implemented using backspace chars and full rewrite of the line containing the progress bar. You probably want to capture the full output of your comand, and make sure that whatever code you have that prints the output, handles the backspaces correctly. See e.g. https://stackoverflow.com/questions/9667462/backspace-behavior-in-python-statement-what-is-correct-behavior-of-printing-a

**Reply by sla-te** [reply]
Finally getting to reply to this one.

Thank you for the hint about backspace-chars. I take it, that in general the suggested implementation for a command, that yields a progress bar is using `_fg=True` then.

It gets difficult if you are calling a complex process like `make`  or `python -m pip install some package` on a remote server and want to log the remote servers stdout to a local file, as `fg=True` disables all that functionality including error handling. The core problem with regards to my question is that such complex commands will yield progress bars as well as regular logs to stdout.  - A more severe problem, that would be caused by using `fg_=True` is, that we would not be able to validate, that the command ran through successfully, which kind of eliminates the core motivation for using a library to wrap commands in if you ask me.

**Reply by ecederstrand** [reply]
Progress bars and logging are not a good match. In general, all logging implementations expect to go forward only, but progress bars don't. Some tools will have a `--batch` flag or similar which prevents emmission of backspaces, color codes etc. But that's up to the tool, of course.

`_fg=True` still causes sh to raise an exception if the executable throws a non-zero exit code.

**Reply by sla-te** [reply]
I agree progress bars and logging are not a good match but its better to have a log with some weird chars in it, than having nothing during running a large command, that runs a lot of sub commands. Its easy to handle this problem ourselves though - but not if we cannot receive stdout or stderr anymore, hehe.

Well, that we are getting the return code Exception is at least something but not really sufficient for a big command, you want specifically detailed logs about this process, else it gets very hard to detect errors.
Say you are calling a build process for your application, then you want to have logs about what `make` (or whatever tool) is doing else it gets kinda scary if you automate deploying the builds to production.

My motivation here was, that it is quite annoying if you copy a large file from a to b and only get any info about what is going on after the command finished.

**Reply by ecederstrand** [reply]
It's not clear to me what you're asking for now. It seems you provided a solution to your own problem in the first post? There's also `_iter=True` that turns the command into a generator of output. That returns the progress bar updates as well, in my experience.

**Reply by sla-te** [reply]
My solution is not an actual solution but rather an unstable hack (`/s in str` would of course not work if you have something like `\some\path` in a stdout message; line break after `/s`) - I was asking if sh has a built-in way to handle it, so I dont have to use it. 

Could you provide an example that copies a file from local to a remote server using scp with `_iter=True`?

**Reply by ecederstrand** [reply]
Your example looks for `/s`, not `\s`. It's not clear to me what you're trying to achieve there, since neither are a Python escape sequence.

**Reply by sla-te** [reply]
You are right, that's a typo in my previous comment, fixed. What I am doing there simply came from testing - I had noticed while running scp, that the progress bar has finished sending an "update" or replacement string for the current values of the progress bar (or rather `aggregated_stdout`) contain `\s` (and I did not deeply research why it does that and what the logic behind it is). My goal though is to move away from my hacky and not thoroughly thought through approach to a best-practise one, so if you could provide an example on how to do it properly, e.g. with `_iter=True`, I could test it and we have a solution if it works (because I did not get that to work). Should be fairly easy for you, as you are probably much deeper into the librarys logic, than I am.

So to reclarify, the example should be a snippet to do the following:
Use scp to transfer a local file to a remote server using sh with the progress bar displayed while not using `_fg=True`.

**Reply by ecederstrand** [reply]
I'm not going to do your work for you, especially since there's still no indication of a bug in sh. There's documentation on `_iter` at https://sh.readthedocs.io/en/latest/sections/special_arguments.html#iter. You can give it a try on your use case.

**Reply by sla-te** [reply]
I think you are exaggerating massively, to my understanding and with regards to my other discussions I am rather heavily contributing to making the library more accessible for other users and am providing very helpful examples and questions, that will assist others and it should be more than obvious, that I have tried before and am not throwing empty tasks into discussions.

Anyway you are probably not the right person to speak as you are not the maintainer, so no hard feelings.

**Reply by ecederstrand** [reply]
I am in fact co-maintainer together with @amoffat. I'm happy for your contributions. I just don't have time to play around with ssh, scp and progress bars myself right now.

**Reply by sla-te** [reply]
Its perfectly fine if you don't have time to do it, hell it would even be ok if you said you don't care or don't want to. Conveying, that I am dropping of work for you to do for me gets me salty though, because I always try to open discussions or tickets in a way, that they will be useful for everyone, not only myself.

Look, I am not asking you to do anything, its an open source library - either you are interested in having more people using it and enriching it by helpful snippets others can use to start of or not. There are explanations for most commands in the docs but there are no examples for any actual use case whatsoever, which makes the approach very time consuming (and I believe I have already digged quite deep).

My approach was to post actual questions with actual approaches and snippets, instead of opening a discussion with the title "add more examples". If that's not sufficient, then I don't know.

**Reply by amoffat** [reply, marked as answer]
>either you are interested in having more people using it ... or not.

@sla-te This is the second time you've expressed that us convincing you to use sh by helping you with your non-sh problems is actually aligned with our interests. It's not, and frankly we couldn't care less if you decided to use sh.

Unfortunately you have used up the remaining good will to help with whatever you are building. If you'd like to earn some of that good will back, you can make a PR with documentation improvements for the things that **you** figure out and would like to share with others. A good starting place might be a tutorial in the [Tutorials Section](https://sh.readthedocs.io/en/latest/tutorials.html)

In the mean time, please don't make any more discussions.
